CMakeLists.txt

```diff
@@ -0,0 +1,106 @@
+cmake_minimum_required(VERSION 3.1)
+
+project(imgui)
+set(IMGUI_VERSION "1.62")
+
+option(IMGUI_BUILD_EXAMPLES "Build imgui examples" OFF)
+
+# Introduce variables:
+# * CMAKE_INSTALL_LIBDIR
+# * CMAKE_INSTALL_BINDIR
+# * CMAKE_INSTALL_INCLUDEDIR
+include(GNUInstallDirs)
+set(IMGUI_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+
+set(IMGUI_HEADERS
+    imconfig.h
+    imgui.h
+    imgui_internal.h
+    stb_rect_pack.h
+    stb_textedit.h
+    stb_truetype.h
+)
+
+set(IMGUI_SOURCE
+    imgui.cpp
+    imgui_demo.cpp
+    imgui_draw.cpp
+)
+
+add_library(imgui INTERFACE)
+
+target_include_directories(imgui INTERFACE 
+    $<BUILD_INTERFACE:${IMGUI_SOURCE_DIR}>)
+
+set(IMGUI_BUILD_SOURCE)
+set(IMGUI_INTERFACE_SOURCE)
+foreach(SOURCE_FILE ${IMGUI_SOURCE})
+    set(IMGUI_BUILD_SOURCE ${IMGUI_BUILD_SOURCE} ${IMGUI_SOURCE_DIR}/${SOURCE_FILE})
+    set(IMGUI_INTERFACE_SOURCE ${IMGUI_INTERFACE_SOURCE} ${CMAKE_INSTALL_INCLUDEDIR}/${SOURCE_FILE})
+endforeach()
+
+target_sources(imgui INTERFACE 
+    $<BUILD_INTERFACE:${IMGUI_BUILD_SOURCE}>
+    $<INSTALL_INTERFACE:${IMGUI_INTERFACE_SOURCE}>
```

Comment:
What is the point of this trick? As far as I can see it means that every project that will use `imgui` package will build from scratch `imgui*.cpp` files. Would it be better for performance reason to build regular library instead?

Reply:
not really sure, the original project had the source directory so I did this to try to minimize the changes to how the original project was used.

Reply:
As far as I can see there is no technical reason for using sources, they just don't have build code (don't want to maintain?). In such situation using a library is a common solution.

Reply:
As far as I can tell they just want the end user to include the cpp files to their project, rather than turn this into a library. I don't know the reasoning for it other than what you mention (dont want to maintain). It is very odd though as a lot of people seem to use this library. They also have very large source files rather than breaking it up as I think most developers would. I imagine the code base is very hard to get started into and work on. They likely could have also turned it into a header only library but didn't go that route either.

Reply:
My main point is that since you've introduced CMake code already, then you can create regular library instead of using tricks. There will be no difference for end users - they will do `target_link_libraries(... imgui::imgui)` in both cases.

Reply:
True, didn't think of it as a static lib. Dlls are always my first thought and didn't want to go adding dllexport everywhere. I'll mod it to a static lib and update.

Reply:
> didn't want to go adding dllexport everywhere

You can use [CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/latest/variable/CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS.html) to export everything, no need to modify C++ code.

Reply:
As with most of the features in cmake, didnt know the existed.

Reply:
looking at the documentation for that it doesn't really work as dllimport still has to be added in order to actually use the dll. Ill just set it up as a static lib only (and fix the other items here).

Reply:
>  it doesn't really work as dllimport still has to be added in order to actually use the dll

I don't think it's true, can you provide an example?

Reply:
https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html#prop_tgt:WINDOWS_EXPORT_ALL_SYMBOLS

# "__declspec(dllimport) must still be used when compiling against the code in the .dll"

Reply:
never mind, that is for data symbols only